garage/tf/q_functions/discrete_mlp_q_function.py

```diff
@@ -8,52 +8,70 @@
 
 class DiscreteMLPQFunction(QFunction):
     """
-    Discrete MLP Function class.
+    Discrete MLP Q Function.
 
     This class implements a Q-value network. It predicts Q-value based on the
     input state and action. It uses an MLP to fit the function Q(s, a).
+
+    Args:
+        env_spec: environment specification
+        hidden_sizes: A list of numbers of hidden units
+            for all hidden layers.
+        hidden_nonlinearity: An activation shared by all fc layers.
+        output_nonlinearity: An activation used by the output layer.
+        layer_norm: A bool to indicate whether to perform
+            layer normalization or not.
     """
 
     def __init__(self,
                  env_spec,
+                 name="discrete_mlp_q_function",
                  hidden_sizes=(32, 32),
                  hidden_nonlinearity=tf.nn.relu,
                  output_nonlinearity=None,
                  layer_norm=False):
-        """
-        Initialize class with multiple attributes.
-
-        Args:
-            env_spec: environment specification
-            hidden_sizes: A list of numbers of hidden units
-                for all hidden layers.
-            hidden_nonlinearity: An activation shared by all fc layers.
-            output_nonlinearity: An activation used by the output layer.
-            layer_norm: A bool to indicate whether to perform
-                layer normalization or not.
-        """
-        super(DiscreteMLPQFunction, self).__init__()
-
+        self._env_spec = env_spec
         self._action_dim = env_spec.action_space.flat_dim
         self._hidden_sizes = hidden_sizes
         self._hidden_nonlinearity = hidden_nonlinearity
         self._output_nonlinearity = output_nonlinearity
         self._layer_norm = layer_norm
 
+        obs_dim = self._env_spec.observation_space.shape
+
+        with tf.name_scope(name):
+            self.obs_ph = tf.placeholder(
+                tf.float32, (None, ) + obs_dim, name="obs")
+            self.q_val = self.build_net(self.obs_ph, name)
+
     @overrides
-    def build_net(self, name, input):
+    def build_net(self, input_var, name):
         """
         Set up q network based on class attributes.
 
         Args:
+            input_var: Input tf.placeholder to the network.
             name: Network variable scope.
-            input: Input tf.placeholder to the network.
         """
         return mlp(
-            input_var=input,
+            input_var=input_var,
             output_dim=self._action_dim,
             hidden_sizes=self._hidden_sizes,
             name=name,
             hidden_nonlinearity=self._hidden_nonlinearity,
             output_nonlinearity=self._output_nonlinearity,
             layer_normalization=self._layer_norm)
+
+    @overrides
+    def get_qval_sym(self, input_phs):
+        assert len(input_phs) == 1, "Only one input should be given."
+
+        return mlp(
```

Comment:
how do i know this is in the same variable_scope as all the other calls to `build_net()`? this is exactly the same problem that Models tries to solve...

Reply:
you are right, we should have a `tf.variable_scope(name, reuse=False)` in `build_net()` and `tf.variable_scope(name, reuse=True)` in `get_qval_sym()`.

Reply:
if you can do it in <30 mins, why not replace this with a DiscreteMLPModel?

Reply:
Yes that's a good idea.